### DISCUSSION TITLE
nREPL standard compliance test suite

### DISCUSSION THREAD

**Discussion by borkdude** [nrepl/nrepl#273] [Ideas]
Following up on the Twitter discussion [here](https://twitter.com/jackrusher/status/1525092640286560258). 

Both Jack and I experienced the following while making non-standard nREPL server implementations (mostly for SCI-based environments like babashka, nbb, ...). Clients usually had to adapt to the server not being a Clojure JVM server or not being a CLJS REPL hosted via a JVM process.

Optimally every nREPL client (editor) should be able to talk to any nREPL server without making assumptions about the language, runtime, etc. The nREPL protocol should be self-sufficient. If a client decides to do stuff "out of band" by probing the server through `eval`, etc, the client should do so cautiously and not degrade the developer experience by doing so.

If this is too much to ask for existing implementations, then there could be a middle ground: every client (editor) should have at least one `general` nREPL target which _only_ plays by the rules of the nREPL protocol.

Instead of documentation, or in addition to it, I think it would be good to have a nREPL server which acts as a standard test so clients can verify if they are meeting the nREPL standard and aren't doing anything that assumes too much about the server.

I would be happy to whip up a such an nREPL standard testing server which will just send back mock data to the nREPL client.

**Comment by jackrusher** [comment]
In addition to what @borkdude said above, I would like that whatever baseline we decide for the protocol to be able to provide _at least_ the fundamentals of a decent browser-connected REPL experience (completions, arglists, docstrings, and so on).

**Reply by borkdude** [reply]
Although the standard should certainly support this, the fact that the nREPL server is hosted in a browser should be pretty much irrelevant?

**Reply by bbatsov** [reply]
Indeed. Perhaps we can open a separate topics about the each op that's missing or poorly specified. E.g. what do we need to change with respect to "completions" (https://nrepl.org/nrepl/0.9/usage/misc.html#code-completion) today? What do we need to change with respect to "lookup" (https://nrepl.org/nrepl/0.9/usage/misc.html#symbol-lookup)?

**Reply by jackrusher** [reply]
It would be nice to get a general consensus about what the ops there should be (that is, including hearing from everyone what they're resorting to `eval` to accomplish now). After that, yeah, let's workshop each op!

**Reply by borkdude** [reply]
Well, let's start with

- clone (start new session)
- describe (say what ops you support (clone, describe at minimum, plus more like eval).
   **clients should not use any other ops than the ones declared here** as supported by the nREPL server
- eval

Those are usually the first I implement in a new server and makes or breaks the developer flow. The rest is kind of optional and can be covered later. By explicitly not supporting ops by not mentioning them in describe, the client should not break by trying those ops and getting no reply, unless the nREPL standard specifies that a request should always get a reply from the server and what this reply should look like in the case of unsupported.

**Comment by tonsky** [comment]
Agree 100% on the premise here. I wrote up my ideas on replacing nREPL here. Not sure if incremental improvement from nREPL is possible but some ideas might be useful https://tonsky.me/blog/network-eval/

**Reply by borkdude** [reply]
Thanks for the write-up, this is valuable feedback.

Bencode vs JSON: not a big issue, but it would be more approachable, I agree. Basing the protocol on bencode does make it truly language agnostic though, since in a new language there might be no JSON library but implementing bencode from scratch is a matter of hours.

The "no upgrade" like you describe in your blog is the main problem we're trying to address here: the protocol should behave like a true protocol and not just some serialization format.

I don't see how nREPL stops you from building parallel eval in a server.

Execution time: could be specified in the nREPL spec as some optional key perhaps.

**Reply by borkdude** [reply]
Btw, I am not looking to replace nREPL. Clients just should behave better and treat nREPL as a protocol and I think just a little is needed to change that. Also the spec needs to be improved. Let's not invent a new one.

https://twitter.com/jsch4efer/status/1525153239234142208

**Reply by tonsky** [reply]
> I don't see how nREPL stops you from building parallel eval in a server.

Good point. Protocol does allow that, but the way sessions/evals work in practice in current nREPL don’t. The question is do we treat de-facto behavior as the part of the standard or not?

“Want parallel eval? Use our nREPL server instead of default one!” Sounds okay to me.

Stacktrace/auto-require/disable stdout capture/eval time could probably be added to nREPL today as options, that would make it much more useful!

**Reply by jackrusher** [reply]
Mostly agree with everything here, @tonsky. So, I'll just add some notes about the points of difference:

Completely neutral on bencode vs JSON, and I say that as someone would directly benefit from being able to send JSON :) 

Evaluation time as an optional piece of the `eval` reply seems reasonable. Making it a requirement, less so.

I think the "Extract info from Clojure file" stuff is the client's problem. I agree it would be nice if there were some shared resource for parsing, &c, on clients, but I don't think it belongs in the protocol. Maybe treesitter is way forward there?

I'm not so into parallel eval as a protocol feature (in the sense of being a promise made by the `eval` op). This is both because there are runtimes where that doesn't make sense (like browsers), and because in the common case you really do want each evaluation to have access to the results of the last evaluation when it starts.

OTOH, I would support the creation of a small library on the runtime side that gives access to operations like start/list/stop for threaded tasks -- analogous to a shell with `ps`, `kill`, and such -- that has persistent re-connectable sessions for doing long-running things. (I *ahem* may or may not have built a private version of this for remote server stuff.)

**Reply by tonsky** [reply]
Re: parallel eval. On client, you control when to run what. If you want one eval to have access to results of another, you wait for it. No?

Weird thing about it is that it doesn’t really require anything special from the protocol. id is already there on both result and response, interrupt is there. The only thing that stops it is default nREPL implementation and session management. Maybe @borkdude is right and it’s a matter of writing alternative server. But I would like to see the support in the vanilla too.

**Reply by mauricioszabo** [reply]
About parallel eval - sometimes it gets tricky. When one refreshes, or load a file, sometimes the runtime behaves *incredibly weird*. I tried in the past, not sure if this is solved, but I remember getting in trouble in the past...

Auto-require is a big no-go for me. I _really don't want_ my `eval` to do things I didn't ask it for

"No upgrade" is tricky for ClojureScript, but I think it's doable if we add more options to "eval" or have a special "Clojure Eval" for cases when it's needed. In my experience with non-trivial ClojureScript code, it's not that easy to ignore that Clojure exists when evaluating ClojureScript, specially when metadata, doc, is needed for something that may be a macro. But I agree 100% that piggieback is a pain, and upgrading the REPL is even worse, so we need a better API indeed.

And I'm not sure what you meant by "Capturing `*out*`"...

**Reply by bbatsov** [reply]
A few quick thoughts.

I liked @tonsky's article, but I definitely think that a few of the concerns there are just implementation details and not protocol issues - e.g. the lack of parallel evals and the concept of "upgrading" a connection. Obviously if you have a native ClojureScript nREPL server or just a server that knows about ClojureScript natively none of the upgrade sequence would be needed.

Same with middleware, btw - they are an implementation detail of the reference Clojure implementation, not something that's part of the nREPL protocol. To my knowledge most nREPL implementation don't have middleware at all and that's totally fine. The JSON support by default is to some extent a related problem - it's kind of hard to implement this without resorting to some third-party library, as JSON is much more complex than bencode. There were also some historical issues that are now addressed - e.g. 10 years ago the JSON parser options available with Emacs and Vim were abysmally slow. I do understand, however, that supporting JSON natively would be beneficial to many client authors today.

I like a lot the idea of automated session management and improved handling of exceptions. Adding some "elapsed time" to evaluations should be pretty simple as well. I think the main reason we didn't do it so far was that some clients might prefer the time it took on the client end to do some evaluation and they don't care much how much time it took on the server (this factors in the network latency + the encoding/decoding).

Like @jackrusher I think that some things are the client's responsibility - e.g. the auto-require, which is actually more nuanced than it seems. Often evals will break because the `ns` form you've last evaled is not in sync with what's currently in your code. CIDER for instances keeps track of this and auto-evals the ns if gets out of sync. I'm open to be persuaded it shouldn't be like this, though. :-)

To be clear - I'm more than happy to discuss and make improvements to the Clojure nREPL server's implementation as well. I just wanted to highlight the difference between the protocol and its implementations.

**Comment by mauricioszabo** [comment]
I would very much like to return a structured value instead of a string. Some interesting tooling could be built around it, like for example inspection of elements, etc.

If this is not possible, at least it would be useful to have a better way of customizing outputs. Currently, it relies on a global var being defined somewhere that you need to pass as an argument to eval - maybe pass the that function inline could be better?

**Reply by tonsky** [reply]
I guess the problem with structured output is semantic mismatch between bencode and EDN. EDN has vectors, maps, sets, dates etc, bencode only has lists, strings and numbers. You’d lose a lot of information that might be important for Clojure consumer

**Reply by mauricioszabo** [reply]
Sure, but one could use `{"type" "vector", "value" [{"type" "keyword", "value" ":foo/bar"}]}` for example...

**Reply by mauricioszabo** [reply]
Or, I don't know, just return transit, or something :D

**Reply by borkdude** [reply]
In general it is not possible to treat evaluation results as structured output, since there might be unserializable objects in there. I think for inspection of elements you could use `eval` for inspection and then do something with that output _on the server side_. E.g. `tap>` etc trigger the side effect of inspecting something, e.g. for displaying in REBL or Portal.

**Reply by mauricioszabo** [reply]
Sure, but the runtime _knows_ that an object is unserializable, so it could just return `#object "Some@JavaThing"`, for example. The client-side doesn't, and malformed EDN happens _a lot_ - like, `{:a (keyword "foo bar")}`, or something.

It's better to get `{:a-object <weird-thing>}` than a string that I can't split into what's EDN and what is not...

**Reply by tonsky** [reply]
```
user=> (pr-str (keyword "foo bar"))
":foo bar"
```

OMG I didn’t know EDN serialization is not reliable! Should this be addressed by Clojure Core though?

**Reply by borkdude** [reply]
This is a long known issue and won't get fixed. Note that serializing random objects using `pr-str` isn't the same as EDN.

**Reply by tonsky** [reply]
What is the proper way to serialize to EDN?

**Reply by borkdude** [reply]
There is no way to serialize `(keyword "foo bar")` to proper EDN. Transit might work better for that. Perhaps the `pr-str` function that gets used in nREPL should be configurable so the client can decode the stringified payload using a custom function. I'm not sure if that's already possible.

**Reply by mauricioszabo** [reply]
@borkdude it is, but we have to pass a global var to make it work. It's doable, but not that straighforward and I suppose people will not like that connecting to a REPL causes a side-effect...

**Reply by jackrusher** [reply]
@mauricioszabo I, for one, dislike that connecting to the REPL changes print behavior (this is the case with current Cider middleware, and caused us some headaches recently in Clerk).

As for structured output, I don't think it makes any sense to require it in the protocol. There are too many weird edge cases and unpredictable things there given the non-round-trippable nature of Clojure's printer/reader. 🤷🏻‍♂️

**Comment by mauricioszabo** [comment]
About ClojureScript: from what I understand, nREPL on it would be a little bit tricky. Currently I'm interfacing directly with shadow-cljs remote API, and it's a *dream come true* to use it - it shows errors when compilation occurs, allows to listen to `tap>` events, etc.

To have a "proper" nREPL on ClojureScript, some extensions would be required a think:
1. Eval - when there's no Javascript environment connected, what do do?
2. Eval - when there are MULTIPLE JS environments connected, what to do by default, and how to control which env it'll run?
3. ClojureScript depends on Clojure code sometimes, in case of macros, for example. How to eval these, considering the nREPL will be on "ClojureScript mode"?
4. Similarly, if we eval Clojure code, we can break the eval, but we can't do that in ClojureScript code. How to handle these differences?
5. If we're running ClojureScript code and the runtime goes away (node crashes, browser gets refreshed), what do we do? Specially considering that parallel eval is not possible in ClojureScript...
6. Can we augment the nREPL protocol to have compilation errors and warnings? These are available on Shadow remote API, and they are incredibly useful
7. Stacktraces in ClojureScript are tricky. Currently nREPL does not send it anyway, but maybe it's interesting to add them?

Probably more to come :D

**Reply by jackrusher** [reply]
My interest in this question stems from my having just built a native nREPL server into a browser that uses @borkdude's SCI interpreter (which might get swapped for self-hosted ClojureScript) to evaluate code without any JSM process involved at all:

https://twitter.com/jackrusher/status/1523706807260114944

**Comment by borkdude** [comment]
I notice several reactions go into the direction of: what more can we do with nREPL?

I'd like to approach this from the side of: let's make the existing clients work as it should: treating the protocol as a protocol without any other assumptions. Not by inventing more features, unless they are strictly necessary to support gaps in the protocol for which now hacks are used.

Feel free to open new discussions with "what more can we do" but let's try to keep it focused here on nREPL spec conformance.

**Reply by tonsky** [reply]
Re-reading the post now, I see that I misunderstood your intentions. I opened separate discussion https://github.com/nrepl/nrepl/discussions/275

**Reply by bbatsov** [reply]
@tonsky Good idea. I'll comment on your suggestions there in more details when I can.

@borkdude I guess it will also be nice to assemble some best practices for client authors, so new clients can avoid some of the problems of the past. I wrote https://nrepl.org/nrepl/0.9/building_clients.html a while ago, but this can be expanded a lot.

**Reply by jackrusher** [reply]
I don't think these are entirely separate questions, @borkdude. If the popular clients can't give good UX using what's in the protocol, they're going to try to route around it with `eval` hacks. I would like to make sure that the protocol does enough that using the ops supported in `describe` doesn't lead to tool authors receiving an endless series of bug reports and feature requests that ultimately incentives them to do the wrong thing.

**Reply by borkdude** [reply]
> I don't think these are entirely separate questions

I already agreed with you on that by saying:

> Not by inventing more features, unless they are strictly necessary to support gaps in the protocol for which now hacks are used.

**Comment by cursive-ghost** [comment]
Cursive deliberately doesn't rely on more than eval and load-file when using nREPL, primarily so that more of the REPL client code can be shared across different REPL types (nREPL, socket REPL, direct IDE eval, debug REPL etc). I don't want my nREPL code to work in a totally different way to other clients. Philosophically I also don't like middleware, I realise that ship has sailed but I'd still rather just rely on the minimum base level of functionality. IntelliJ's autocomplete functionality also doesn't work well with an RPC server since it's invoked a lot, IntelliJ will generally pop up autocompletions without explicit user action and updates them frequently on typing. So the nREPL completion op isn't much use to me anyway.

Cursive tries to detect what sort of REPL it's connecting to (CLJ, CLJS, Babaska etc), and then falls back to JVM Clojure if it can't. Jack mentioned in the Twitter thread that this clearly doesn't work for new server types, which is a fair point. I'll update the Cursive client to fall back to something non-JVM-specific, which will just autocomplete Clojure code based on standard introspection and not do anything for interop forms.

Actually, looking briefly at my code, I do use a couple more ops: interrupt and stdin. I can easily toggle whether interruption is available based on describe, but I'm not sure how I would toggle stdin - I guess if the server never sends needs-input then I'll never use it.

**Reply by borkdude** [reply]
> I'll update the Cursive client to fall back to something non-JVM-specific

Excellent. Then that should work for nbb and Jack's Hubble.

> which will just autocomplete Clojure code based on standard introspection and not do anything for interop forms

That's unfortunate since if the nREPL server implements the completion op, it can be quite useful. E.g. Calva + nbb , you get auto-completions for JS interop. I don't know what makes Calva or Cursive so different that the nREPL completion op doesn't work for Cursive.

See e.g. https://twitter.com/borkdude/status/1504488345681870858

> I can easily toggle whether interruption is available based on describe

Good! That is how describe should be used.

Do you agree that a nREPL testing server against which you can test your nREPL client (Cursive) could be useful to test if you meet a certain degree of compliance?

**Reply by cursive-ghost** [reply]
Sure, that might be useful. But I think the only thing I'm doing that violates the protocol is calling ops without checking first with describe that they're available. Sending JVM code to non-JVM servers isn't a protocol violation as such, although I admit it's not very helpful :)

**Reply by cursive-ghost** [reply]
> I don't know what makes Calva or Cursive so different that the nREPL completion op doesn't work for Cursive.

Does Calva pop up autocompletions without user interaction (i.e. you don't have to actively do anything to see completions), and then update the completions on every keystroke? I wrote all this code a long time ago, but when I tried back then on e.g. connections to remote machines that was painfully slow.

**Reply by borkdude** [reply]
@cursive-ide It's very easy to try this for yourself. 

In a terminal, install nbb:

```
$ npm install -g nbb
```

Then open Calva in a directory with an `example.cljs` file:

``` clojure
(ns example
  (:require ["fs" :as fs]))
```

```
$ code .
```

Cmd-Shift-P: Calva: connect to a Clojure REPL, Jack-in, select nbb

Then evaluate the ns form. And start typing `fs/`

**Reply by borkdude** [reply]
It does seem the completions appear automatically when typing (you don't need to hit tab, if that is what you mean). It's pretty fast.

> but when I tried back then on e.g. connections to remote machines that was painfully slow.

Isn't that expected, with a network in the middle, things get usually slower?

**Reply by jackrusher** [reply]
Cider also updates completions and docstrings on every keystroke with no bother.

One reason I would prefer that Cursive use the protocol is that the live runtime knows things that static analysis absolutely cannot. For example, the Hubble server provides completions, arglists, and docstrings for JS functions (both browser built-ins and ES6 modules pulled in dynamically over the network).

Would it maybe make sense to use the normal completion stuff but also have an option to turn off those completions in case someone is eval'ing on a remote machine via a 1200 baud modem?

**Reply by cursive-ghost** [reply]
Ok, I'll give it a go. Thinking about it, I restructured my REPL code a while back, so it actually might be easier than it was to use the nREPL ops just for the nREPL client, I'll check.

> Isn't that expected, with a network in the middle, things get usually slower?

Well, yes, but the point is that Cursive doesn't get slower because it doesn't use the network in such a chatty way.

> One reason I would prefer that Cursive use the protocol is that the live runtime knows things that static analysis absolutely cannot. For example, the Hubble server provides completions, arglists, and docstrings for JS functions (both browser built-ins and ES6 modules pulled in dynamically over the network).

Everything's a tradeoff for sure. One thing to point out here, Cursive doesn't use the REPL at all for completion in the standard code editor. It's only used in the REPL input pane, which is becoming an increasingly niche use case anyway. Integrating a combination of static analysis and REPL results in the completions is something that gets requested from time to time, but would be a lot of work to do well. I'm actually planning some work to make the completion popup more responsive, and it might be easier at that point to integrate async completion results into the final list. But there are lots of UI issues too, Cursive allows the user to have multiple REPL sessions open, something I frequently do. So the user would have to be able to specify which REPL completions should come from, among other problems.

**Reply by imrekoszo** [reply]
Would it make sense for the server to serve a database of completions that a client could use to cache runtime info in case live updates affect perf too much?

**Reply by cursive-ghost** [reply]
FWIW, that's more or less what Cursive does right now. At startup it figures out the current namespace, and downloads the completion info for vars etc available in that namespace, and also the info for namespaces available via aliases etc. If any eval results in a change of namespace, Cursive updates its local data with that namespace and its dependencies.

**Reply by borkdude** [reply]
That's a server implementation detail that doesn't have anything to do with the nREPL protocol compliance we're focusing on here, I think?

**Reply by imrekoszo** [reply]
> server implementation detail ... nREPL protocol compliance we're focusing on here, I think?

Sure I suppose, just wanted to mention it as the question "why certain clients don't use the current completion api" came up.

Edit: reading back a bit this seems to relate to @jackrusher 's comment at https://github.com/nrepl/nrepl/discussions/273#discussioncomment-2750063

**Reply by cursive-ghost** [reply]
I don't think it's a server implementation detail, it's a different type of completion API, one which provides all the information up front rather than on-demand.
